evm/src/arithmetic/mul.rs

```diff
@@ -10,23 +10,49 @@
 //!
 //!    A = \sum_{i=0}^15 a[i] β^i,
 //!
-//! where β = 2^16. To verify that A, B and C satisfy the equation we
-//! proceed as follows. Define a(x) = \sum_{i=0}^15 a[i] x^i (so A = a(β))
-//! and similarly for b(x) and c(x). Then A*B = C (mod 2^256) if and only
-//! if there exist polynomials q and m such that
+//! where β = 2^16 = 2^LIMB_BITS. To verify that A, B and C satisfy
+//! the equation we proceed as follows. Define
 //!
-//!    a(x)*b(x) - c(x) - m(x)*x^16 - (β - x)*q(x) == 0.
+//!    a(x) = \sum_{i=0}^15 a[i] x^i
+//!
+//! (so A = a(β)) and similarly for b(x) and c(x). Then A*B = C (mod
+//! 2^256) if and only if there exists q such that the polynomial
+//!
+//!    a(x) * b(x) - c(x) - m(x) * q(x)
+//!
+//! is zero when evaluated at x = β, i.e. it is divisible by (x - β);
+//! equivalently, exists a polynomial s such that
+//!
+//!    operation(a(x), b(x)) - c(x) - m(x) * q(x) - (x - β) * s(x) == 0
+//!
+//! if and only if operation(A,B) = C (mod M). In the code below, this
+//! "constraint polynomial" is constructed in the variable
+//! `constr_poly`. It must be identically zero for the multiplication
+//! operation to be verified, or, equivalently, each of its
+//! coefficients must be zero. The variable names of the constituent
+//! polynomials are (writing N for N_LIMBS=16):
+//!
+//!   a(x) = \sum_{i=0}^{N-1} input0[i] * β^i
+//!   b(x) = \sum_{i=0}^{N-1} input1[i] * β^i
+//!   c(x) = \sum_{i=0}^{N-1} output[i] * β^i
+//!   m(x) = \sum_{i=0}^{N-1} modulus[i] * β^i
+//!   q(x) = \sum_{i=0}^{N-2} quot[i] * β^i
+//!   s(x) = \sum_i^{2N-3} aux[i] * β^i
 //!
 //! Because A, B and C are 256-bit numbers, the degrees of a, b and c
-//! are (at most) 15. Thus deg(a*b) <= 30, so deg(m) <= 14 and deg(q)
+//! are (at most) 15. Thus deg(a*b) <= 30, so deg(q) <= 14 and deg(s)
 //! <= 29. However, the fact that we're verifying the equality modulo
 //! 2^256 means that we can ignore terms of degree >= 16, since for
 //! them evaluating at β gives a factor of β^16 = 2^256 which is 0.
 //!
-//! Hence, to verify the equality, we don't need m(x) at all, and we
-//! only need to know q(x) up to degree 14 (so that (β - x)*q(x) has
-//! degree 15). On the other hand, the coefficients of q(x) can be as
-//! large as 16*(β-2) or 20 bits.
+//! Hence, to verify the equality, *we don't need q(x) (quot) at all*,
+//! and we only need to know s(x) up to degree 14 (so that (x - β)*s(x)
+//! has degree 15). On the other hand, the coefficients of s(x) can be
+//! as large as 16*(β-2) or 20 bits.
+//!
+//! Note that, unlike for the general modular multiplication (see the
+//! file `modular.rs`), we don't need to check that output is reduced,
+//! since any value of output is less than x^16 and is hence reduced.
```

Comment:
Maybe this would be more easily explained as classic long multiplication? Because that’s exactly what this method is (`s` are the carries).